.build/ci/ci_parser.py

```diff
@@ -0,0 +1,307 @@
+#!/usr/bin/env python3
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""
+Script to take an arbitrary root directory of subdirectories of junit output format and create a summary .html file
+with their results.
+"""
+
+import argparse
+import cProfile
+import pstats
+import os
+import shutil
+import xml.etree.ElementTree as ET
+from typing import Callable, Dict, Tuple, Type
+from pathlib import Path
+
+from junit_helpers import JUnitResultBuilder, JUnitTestCase, JUnitTestSuite, JUnitTestStatus, LOG_FILE_NAME
```

Comment:
Have we considered using Jenkins plugins instead of custom scripts to format the results?

According to the INFRA page, the Jenkins JUnit Plugin is installed for the ci-cassandra.
https://cwiki.apache.org/confluence/display/INFRA/ci-cassandra.apache.org

This plugin natively supports JUnit XML report files to print the report:
https://plugins.jenkins.io/junit/

Reply:
I see it was used in the `Jenkinsfile` before, but it didn't suit our needs, did it?

```
junit testResults: '**/build/test/**/TEST*.xml,**/cqlshlib.xml,**/nosetests.xml', testDataPublishers: [[$class: 'StabilityTestDataPublisher']]
```

Reply:
`junit testResults` is used in the Jenkinsfile to provide the "Test Result" UI.
e.g. https://ci-cassandra.apache.org/view/patches/job/Cassandra-5-devbranch/92/testReport/

while this collects our many thousands of xml unit test files, it provides no way of aggregating them into one, nor of exporting the results (either whole or per test-type) in aggregated xml unit file. 

the structure and content of the ci_summary.html and results_details tarball is becoming standardised across this and apple's CI system.  it requires aggregated unit results (single xml and folder of html) per test type.

the jenkins UI is not always available.  i.e. we could be running this Jenkinsfile in a private jenkins instance, and the ci_summary.html and results_details tarball are the only two files we have to then share.  in [CASSANDRA-18145](https://issues.apache.org/jira/browse/CASSANDRA-18145) we'll provide `.build/run-ci.sh …` script that setups up in k8s a fresh jenkins, runs the pipeline, gets these two files, tears down the jenkins for you, so that you never have to open the jenkins UI at all (everything on the command line).

also note, aggregating all our test results together requires >14GB ram (all plugins use xerces which isn't streaming). 

a lot of these requirements and challenges are written up in [CASSANDRA-18137](https://issues.apache.org/jira/browse/CASSANDRA-18137)